src/H5T.c

```diff
@@ -5245,67 +5271,245 @@ H5T__path_find_real(const H5T_t *src, const H5T_t *dst, const char *name, H5T_co
             HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "unable to copy datatype for conversion path");
     } /* end if */
     else
-        path = table;
+        path = matched_path;
+
+    /* Initialize the path if it's a new path */
+    if (new_path && H5T__path_find_init_new_path(path, src, dst, conv, &tmp_ctx) < 0)
```

Comment:
Moved the two large sections that initialize a new path with either the specified hard conversion function or an existing soft function into the new `H5T__path_find_init_new_path` helper function, making the purpose of those sections clearer and localizing the temporary datatypes and IDs for those types used when initializing the conversion function